### PR TITLE
Security enhancements for data exhaust APIs

### PR DESCRIPTION
**Pull request by sowmya-dixit**

### CODE REVIEW

File: analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/auth_verifier/AccessTokenValidator.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.ekstep.analytics.api.util.JSONUtils
+import java.util.Base64
+
+import javax.inject.Singleton
+import org.ekstep.analytics.framework.conf.AppConf
+
+@Singleton
+class AccessTokenValidator {
+
+    implicit val className = "org.ekstep.analytics.api.util.auth_verifier.AccessTokenValidator"
```
Comment (SanthoshVasabhaktula) [inline analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/auth_verifier/AccessTokenValidator.scala:13]: Fix the package name. remove "auth_verifier' from the package name

File: analytics-api/app/controllers/JobController.scala [hunk 2]
```diff
@@ -116,15 +123,68 @@ class JobController @Inject() (
...
+        }
         else {
-            val status = Option(cacheUtil.getConsumerChannelTable().get(consumerId, channelId))
-            if (status.getOrElse(0) == 1) (true, None) else (false, Option(s"Given X-Consumer-ID='$consumerId' and X-Channel-ID='$channelId' are not authorized"))
+            // get userId from user auth token
+            val userId = accessTokenValidator.getUserId(userAuthToken.get)
+            var unauthorizedErrMsg = "You are not authorized."
+            APILogger.log("userId retrieved: " + userId)
+            if(!"Unauthorized".equalsIgnoreCase(userId)) {
+                val headers = Map("x-authenticated-user-token" -> userAuthToken.get, "Authorization" -> authBearerToken.getOrElse(""))
```
Comment (SanthoshVasabhaktula) [inline analytics-api/app/controllers/JobController.scala:151]: i don't think we need to send the bearer token if we are calling the service via internal end point

File: analytics-api/app/controllers/JobController.scala [hunk 3]
```diff
@@ -116,15 +123,68 @@ class JobController @Inject() (
...
+                                val orgSearchApiUrl = config.getString("org.search.url")
+                                val requestBody = """{"request":{"filters":{"channel":"mhrd"},"offset":0,"limit":1000,"fields":["id"]}}"""
+                                val response = restUtil.post[Response](orgSearchApiUrl, requestBody)
+                                APILogger.log("org search response: " + JSONUtils.serialize(response))
+                                val contents = response.result.getOrElse(Map()).getOrElse("response", Map()).asInstanceOf[Map[String, AnyRef]]
+                                  .getOrElse("content", List(Map())).asInstanceOf[List[Map[String, AnyRef]]]
+                                val mhrdChannel = if(contents.size > 0) contents.head.getOrElse("id", "").asInstanceOf[String] else ""
+                                val userChannel = orgDetails.getOrElse("channel", "").asInstanceOf[String]
+                                APILogger.log("user channel: " + userChannel + " mhrd id: " + mhrdChannel)
+                                if (userChannel.equalsIgnoreCase(mhrdChannel)) return (true, None)
```
Comment (SanthoshVasabhaktula) [inline analytics-api/app/controllers/JobController.scala:175]: Super admin channel needs to be configured and this needs to be cached on service start

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/auth_verifier/AccessTokenValidator.scala (hunk 1 above)
Comment (sowmya-dixit) [reply analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/auth_verifier/AccessTokenValidator.scala]: Done removed package name

File: analytics-api/app/controllers/JobController.scala (hunk 3 above)
Comment (sowmya-dixit) [reply analytics-api/app/controllers/JobController.scala]: Added super admin channel to cache on service start

File: analytics-api/app/controllers/JobController.scala (hunk 2 above)
Comment (sowmya-dixit) [reply analytics-api/app/controllers/JobController.scala:150]: For these changes give new PR

Comment (sowmya-dixit) [reply analytics-api/app/controllers/JobController.scala]: Done - removed bearer token as we are using internal end point

File: analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/CacheUtil.scala [hunk 4]
```diff
@@ -29,12 +26,32 @@ case class LanguageResponse(id: String, ver: String, ts: String, params: Params,
...
+    initDeviceLocationCache()
+    initConsumerChannelCache()
+    initSuperAdminChannelCache()
+  }
+
+  def initSuperAdminChannelCache()(implicit config: Config) {
+    APILogger.log("Updating super admin channel cache ")
+    // get MHRD tenant id using org search API
+    val orgSearchApiUrl = config.getString("org.search.url")
+    val requestBody = """{"request":{"filters":{"channel":"mhrd"},"offset":0,"limit":1000,"fields":["id"]}}"""
```
Comment (SanthoshVasabhaktula) [inline analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/CacheUtil.scala:47]: {"channel":"mhrd"} - "mhrd" value should config from config as it is diksha specific

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/CacheUtil.scala (hunk 4 above)
Comment (sowmya-dixit) [reply analytics-api-core/src/main/scala/org/ekstep/analytics/api/util/CacheUtil.scala]: Done - added channel as config and defaulted to sunbird

**Review by SanthoshVasabhaktula — APPROVED** [review]

**Pull request merged**